### PR TITLE
fma relu combination for convolution-output

### PR DESCRIPTION
**Pull request by Jokeren**
Thanks for your review!

### CODE REVIEW

**Comment by Maratyszcza** [comment]
Thanks Keren. Looks good, but needs few changes:

1. Please use `activation` parameter of enum type rather than a `bool`. In the future, we may want to support other activation types, not just ReLU. I.e.

```
enum nnp_activation {
	nnp_activation_identity = 0,
	nnp_activation_relu = 1,
};

enum nnp_status nnp_convolution_output(
	enum nnp_convolution_algorithm algorithm,
	size_t batch_size,
	size_t input_channels,
	size_t output_channels,
	struct nnp_size input_size,
	struct nnp_padding input_padding,
	struct nnp_size kernel_size,
	const float input[],
	const float kernel[],
	const float bias[],
	float output[],
	enum nnp_activation activation,
	pthreadpool_t threadpool,
	struct nnp_profile* profile);
```

2. Please generate separate FFT functions with ReLU, e.g. `nnp_ifft8x8_with_bias_with_relu__avx2` instead of argument for inverse FFT functions. Also, not that if you have `arg_relu` parameter, `if arg_relu:` statement does not generate assembly instructions to check if arg_relu is non-zero, rather it checks if the Python variable `arg_relu` evaluates to `True`.

3. Please use tabs for indentation in C/C++ sources.

File: src/x86_64-fma/fft16x16.py [hunk 1]
```diff
@@ -509,6 +514,8 @@ def inverse_vfft(reg_t0, reg_t8, reg_t_stride, data_in, reg_row_start=None, reg_
                 elif reg_row_end:
                     CMP(reg_row_end, row_lo)
                     JBE(store_data.end)
+                if relu:
+                    VBLENDVPS(ymm_data_lo, ymm_data_lo, ymm_zero, ymm_data_lo)
```
Comment (Maratyszcza) [inline src/x86_64-fma/fft16x16.py]: It is preferable to use `VMAXPS(ymm_data_lo, ymm_data_lo, ymm_zero)` for performance reasons (`VBLENDPS` may generate multiple microoperations)

File: src/x86_64-fma/fft16x16.py [hunk 2]
```diff
@@ -499,6 +500,10 @@ def inverse_vfft(reg_t0, reg_t8, reg_t_stride, data_in, reg_row_start=None, reg_
                     negate_b=fft8_negate_b.get(id(data_hi), False),
                     writeback=False)
 
+            if relu:
+                ymm_zero = YMMRegister()
+                VMOVAPS(ymm_zero, Constant.uint32x8(0))
```
Comment (Maratyszcza) [inline src/x86_64-fma/fft16x16.py]: Please use negative signed zero, i.e. `Constant.float32x8(-0.0)`

**Comment by Jokeren** [comment]
Sorry that I set `tab=8 space` previously so that the style is not consistent with yours. 

I have changed the format, and it looks just fine. 

I will fix other functions later.

Regarding the `psimd` implementations, I do not have a mac machine. Then, how to test them after adding the relu function?

**Comment by Maratyszcza** [comment]
You don't need a Mac to test `psimd` functions, any Linux machine with Clang would work. Just do `python configure.py --enable-psimd`

File: src/x86_64-fma/fft16x16.py (hunk 2 above)
Comment (Jokeren) [inline src/x86_64-fma/fft16x16.py]: Could you please tell me the reason to use negative signed zero? Or propose a simple example?

**Comment by Jokeren** [comment]
Hi, @Maratyszcza !

Please view my latest commits to see whether the structures and format meet your requirement.

A problem about testing is proposed in the issues.

File: bench/vgg.c [hunk 3]
```diff
@@ -94,7 +94,9 @@ double benchmark_vgg(
 				for (size_t layer_index = 0; layer_index < layers_count; layer_index++) {
 					switch (layers[layer_index].type) {
 						case layer_type_convolutional:
-							status = nnp_convolution_output(nnp_convolution_algorithm_auto,
+							status = nnp_convolution_output(
+								nnp_activation_identity,
```
Comment (Maratyszcza) [inline bench/vgg.c]: Is the order correct? In `include/nnpack.h` activation is the second argument

File: src/convolution-output.c [hunk 4]
```diff
@@ -458,14 +455,32 @@ enum nnp_status nnp_convolution_output(
...
-			output_transform_function = nnp_hwinfo.transforms.ifft8x8_with_bias;
+			switch (activation) {
+				case nnp_activation_relu:
+					output_transform_function = nnp_hwinfo.transforms.ifft8x8_with_bias_with_relu;
+					break;
+				case nnp_activation_identity:
+					output_transform_function = nnp_hwinfo.transforms.ifft8x8_with_bias;
+					break;
+				default:
+					goto cleanup;
```
Comment (Maratyszcza) [inline src/convolution-output.c]: `nnp_convolution_output` should return an error code if `activation` has unknown value. I suggest to check activation inside `validate_convolution_arguments`. Then in these switch statements you write `NNP_UNREACHABLE;` to indicate that this case never happens. Compiler will use it for optimization.

File: src/x86_64-fma/fft16x16.py (hunk 2 above)
Comment (Maratyszcza) [inline src/x86_64-fma/fft16x16.py]: This is for the backward pass. The backward pass needs to know which values were positive/negative before we applied ReLU. Using negative zero ensures that the sign of the convolution results doesn't change after we apply ReLU. See discussion in #24 for why its important.

**Comment by Maratyszcza** [comment]
Looks good. Once the tests are working I will merge.

**Comment by Jokeren** [comment]
Hi, @Maratyszcza , all tests are passed!

I am sorry that my former commit logs are not in a correct format.

I used [-0.1, 1] uniform distribution for convolution tests.

**Comment by bhack** [comment]
Any news?

**Comment by Maratyszcza** [comment]
@bhack This PR is ready to merge, but first NNPACK is moving to a new configuration system.

**Comment by Jokeren** [comment]
Anything I can help? @Maratyszcza

**Comment by Maratyszcza** [comment]
@Jokeren There are reports of a bug in `nnp_fully_connected_output` #43 #44. Could you take a look?

**Comment by Maratyszcza** [comment]
Manually rebased, merged, and committed as 4dbf75dd2e4c98461e77d9730744fa101b6de219

**Pull request closed without merging**